Problem.cpp

```diff
@@ -519,29 +519,106 @@ void Problem::print_sources() {
     //Open the CSV file for writing
-    std::ofstream file("./Output/data.csv");
+    std::ofstream file("./Output/valid.csv");
     if (!file.is_open()) {
         std::cerr << "Error opening file to write output." << std::endl;
     }
     
     //Useful shortcuts and settings
     unsigned int nn = problem_mesh->get_number_of_nodes();
-    unsigned int ne = problem_mesh->get_number_of_elements();
     int decimal_places = 14;
     float factor = pow(10, decimal_places);
     unsigned int i_max {0};
-    if (save_solution == false) { 
-        save_valid = false;
+    
+    //Write the configuration as the first line (ROW 0)
+    file 
+    << nn                     << ","
+    << number_elements_length << "," 
+    << number_elements_depth  << ","
+    << domain_length          << ","
+    << domain_depth           << ","
+    << pulse_intensity        << ","
+    << pulse_frequency        << ","
+    << delta_time             << ","
+    << n_steps                << ","
+    << receivers.size()       << ","
+    << sources.size()         << ","
+    << control->get_nlvls();
+    
+    //Write the solution to the file (ROW 1)
+    file << "\n";
+    i_max = 0;
+    for (unsigned int s = 0; s < sources.size(); ++s) {
+        for (unsigned int t = 0; t < n_steps; ++t) {
+            for (unsigned int n = 0; n < nn; ++n) {
+                file << round((*solution)(n,t,s) * factor) / factor;
+                i_max++;
+                if (i_max != sources.size()*n_steps*nn) {
+                    file << ",";
+                }
+            }
+        }
+    }
+    
+    //Write the stiffness matrix (constant matrix) to the file (ROW 2)
+    file << "\n";
+    i_max = 0;
+    for (arma::uword j = 0; j < nn; ++j) {
+        for (arma::uword i = 0; i < nn; ++i) {
+            file << round((*global_stiffness_consistent)(i,j) * factor) / factor;
+            i_max++;
+            if (i_max != nn*nn) {
+                file << ",";
+            }
+        }
     }
-    if (save_valid == true) {
-        save_solution = true;
+    //Write the mass matrix to the file (ROW 3)
+    file << "\n";
+    i_max = 0;
+    for (arma::uword j = 0; j < nn; ++j) {
+        for (arma::uword i = 0; i < nn; ++i) {
+            file << round((*global_mass_consistent)(i,j) * factor) / factor;
+            i_max++;
+            if (i_max != nn*nn) {
+                file << ",";
+            }
+        }
     }
     
+    //Close the file
+    file.close();
+    std::cout << "Done. " << std::endl;
+    
+}
+
+void Problem::write_output(bool save_solution, unsigned int shot_id = 0, unsigned int sample_size = 50) {
+    
+    std::cout << "Writing output to './Output/data.csv'    ...   " << std::flush;
+    std::filesystem::create_directory("./Output/");
+    //Open the CSV file for writing
+    std::ofstream file("./Output/data.csv");
+    if (!file.is_open()) {
+        std::cerr << "Error opening file to write output." << std::endl;
```

Comment:
I believe that if the file couldn't be open, you should return after the message, no?
Something like this:

```
    if (!file.is_open()) {
        std::cerr << "Error opening file to write output." << std::endl;
        return;
    }
```